src/eval.rs

```diff
@@ -197,10 +285,147 @@ impl Expr {
+                if let Ok(x) = index {
+                    if let ExprSource::Dynamic { inner: Ok(val) } = &x.source {
+                        out.push(val.as_ref());
+                    }
+                }
+                if let Ok(x) = from {
+                    out.push(x.as_ref());
+                }
+                return out;
+            }
         }
         .into_iter()
         .map(|x| x.as_ref())
         .filter_map(Result::ok)
+        .map(|x| x.as_ref())
         .collect()
     }
+
+    pub fn get_definition(&self) -> Option<Gc<Expr>> {
+        use ExprSource::*;
+        match &self.source {
+            Ident { name } => self.scope.get(&name),
+            Select { from, index } => {
+                let idx = index.as_ref().ok()?.as_ident().ok()?;
+                let out = from
+                    .as_ref()
+                    .ok()?
+                    .eval()
+                    .ok()?
+                    .as_map()
+                    .ok()?
+                    .get(&idx)?
+                    .clone();
+                if let ExprSource::KeyValuePair { ref key, .. } = out.source {
+                    key.clone().ok()
+                } else {
+                    Some(out)
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Interpret the expression as an identifier. For example:
+    /// ```text
+    /// foo => "foo"
+    /// "foo" => "foo"
+    /// "${"foo"}" => "foo"
+    /// ```
+    pub fn as_ident(&self) -> Result<String, EvalError> {
+        use ExprSource::*;
+        match &self.source {
+            Ident { name } => Ok(name.clone()),
+            Dynamic { inner } => inner.as_ref()?.eval()?.as_str(),
+            Literal { value } => value.as_str(),
+            _ => Err(EvalError::Internal(InternalError::Unimplemented(
+                "unsupported identifier expression".to_string(),
+            ))),
+        }
+    }
+}
+
+/// Used for merging sets during parsing. For example:
+/// { a.b = 1; a.c = 2; } => { a = { b = 1; c = 2; }; }
+pub fn merge_set_literal(name: String, a: Gc<Expr>, b: Gc<Expr>) -> Result<Gc<Expr>, EvalError> {
+    // evaluate literal attr sets only, otherwise error
+    let eval_literal = |src: Gc<Expr>| {
+        let src = if let ExprSource::KeyValuePair { value, .. } = &src.source {
+            value.as_ref()?.clone()
+        } else {
+            src
+        };
+        if let ExprSource::AttrSet { .. } = &src.source {
+            src.eval()?.as_map()
+        } else {
+            // We cannot merge a literal with a non-literal. This error is
+            // caused by incorrect expressions such as:
+            // ```
+            // repl> let x = { y = 1; }; in { a = x; a.z = 2; }
+            // error: attribute 'a.z' at (string):1:33 already defined at (string):1:26
+            // ```
+            Err(EvalError::Value(ValueError::AttrAlreadyDefined(name.to_string())))
```

Comment:
I'm not sure if I understand the issue here, isn't `x` an attr-set as well? (I mean yes, this is an eval-error, but I don't undestand how this is caught by your if-else here).

Reply:
Ah, x is an `ExprSource::Ident`. I've added a comment explaining this, because I agree that the code looks confusing

Reply:
Ahh right, thanks!